### ISSUE TITLE
Concept for app center

### ISSUE THREAD

**Issue by vollstock** [vollstock/Tooloop-Settings-Server#4]
- [ ] Download app bundle (e.g. from github)
- [ ] UI for media
- [ ] How to browse/discover apps?
- [ ] User rating?

**Comment by vollstock** [comment]
We should still be able to upload non-published software.
This should be possible:

1. Client buys/owns a Tooloop box. of a certain model.
2. He gets some software from an artist that works on that model.
3. Client uploads the software and is happy.

This workflow is about money, of course.

**Comment by vollstock** [comment]
# Proposal

## Concept

The app Center will receive a major upgrade. It can not only install presentations but all sorts of bundled apps, services and features. 

In general there are two kinds of bundles: presentations and (system) features. 

A presentation is the running show on this Tooloop box. It typically runs in full screen and has exclusive ownership of the display. There is only one active presentation at a time on a box. 

On the other hand, one can install an arbitrary number of features which can be enabled and disabled at will. Features are useful additions to the systems functionality that not every user or every presentation needs and thus isn’t installed or active per default. Examples for features would be openFrameworks, a transparent mouse cursor, a local web server like nginx, etc.

Presentations may depend on a feature. 
Dependencies will automatically be installed with the app. 
Dependencies will not be uninstalled automatically. 

The optional install script will be replaced entirely by the new app center.

![app center](https://user-images.githubusercontent.com/4962676/45084625-10cfc100-b0ff-11e8-9de7-027201ffef63.png)
![app center presentations](https://user-images.githubusercontent.com/4962676/45084624-10372a80-b0ff-11e8-9438-3a51a1483550.png)
![app center presentation detail](https://user-images.githubusercontent.com/4962676/45084623-10372a80-b0ff-11e8-939a-458d3f005e3f.png)
![app center features](https://user-images.githubusercontent.com/4962676/45084621-10372a80-b0ff-11e8-8a4f-d3db89fdb600.png)



Here is a click dummy of the Settings server containing the proposed app center: https://xd.adobe.com/view/3b79b3bd-ae7b-431a-67f0-10506dae0190-4d48/


## Github workflow

The Tooloop-Examples repo will be replaced by a central app repository that the app center pulls from.

If you want your bundle listed in the app center:

- Open a GitHub repository for your app bundle
- Open an issue on the app repository, asking to add your bundle
- The bundle repo will then be added as a submodule
- Local update of app catalogue is merely a recursive git pull


## Todo

- update bundle definitions
- Features need setting pages of their own (more than one at a time needs to be possible)
- add online status to system controller so we know whether we can update the system and the catalogue and install stuff


## Discussion

- Do we want to moderate bundles or go the ofxAddons approach and crawl github automatically
- (how) does rating/commenting work?
- Mark dependency
- Is this approach bad for future paid bundles in any way?

**Comment by b3nson** [comment]
👍 **nice**

Two thoughts:

- The AppCenter/Presentations-Listing should somehow differentiate clearly between only locally available apps (self-developed/closed-source/testing) and the ones (freely) available on github etc!?
(So, that you don't have to search for your own local app in a growing list of external apps)

- Is "features" the right wording? Enough self-explanatory? As these "features" can be quite everything (frameworks/libraries/runtimes/services/daemons/programs/apps/scripts/whatever) it's not easy to find something suitable. Maybe "addons", "system-addons", "extensions", "box extensions",
superpowers....i don't know 😄 

In general the nextcloud appstore, altough a centralized one, may be worth a look at... (already very similar to yours)
https://github.com/nextcloud/appstore
https://apps.nextcloud.com/


If Tooloop will offer a centralized AppCenter in the future, nextcloud appstore might be a good point to start. :)

**Comment by b3nson** [comment]
Maybe it would be a good idea to separate the global (let's call it) **AppStore** and the local (let's call it) **AppManagement** from each other.

- The **AppManagement** is a local (_offline_) UI (part of Tooloop-Settings-Server) to find, install, configure and uninstall any app that is locally available on the system.

- The **AppStore** is a global/_online_ UI (part of the Tooloop-Website) that aggregates public available apps (adds rating/commenting/versioning/....) and offers them for download. So you can download an app-bundle on machine A and copy it to your box B manually, or launch the **AppStore** and download directly on your box (when you're online). The app then appears in **AppManagement** to be installed/configured etc.

The **AppStore** can of course also be a visually integrated part of Tooloop-Settings-Server-UI (own menu-item / tab). 
But as the **AppStore** is a separate optional service that requires internet connection, and the **AppManagement** is a fundamental local offline tool to administer you box, separation would make things clearer – user-experience- and technical-wise.

?!?😬

**Comment by vollstock** [comment]
@b3nson : sorry, was some days afk and had to do some money work after that :-/
I agree with almost everything you pointed out. In order:

1. I’d like to include a function „Install from file“, that would let you upload a ZIP file with a bundle so you can always install your own proprietary software or a package your coder gave you.
2. Thought about the naming, too. I like „addons“. I also like the Cinder way: „blocks“. Addons is probably the most clear albeit not the most fun naming. 🤔
3. Separation between local app management and online store is great, as that makes the undertaking doable in a manageable effort. This is alpha anyway and we should aim for low hanging fruit!

I have updated the [prototype](https://xd.adobe.com/view/3b79b3bd-ae7b-431a-67f0-10506dae0190-4d48/) accordingly.

**Comment by vollstock** [comment]
I am reading into creating a local repository of .deb packages.
It’s a fairly simply change to how we create packages and read them in the settings server.

The beauty of it is that this way, `apt` will do all the heavy work:
different processor architectures, package versions, installation and most important: dependencies.

So far, this works like a charm here in my test system :-)

It will be quite a ride to turn all optional install scripts into `.deb`s, though 🤔

**Comment by vollstock** [comment]
# Day 2 of hacking my own (local) .deb repository

Successfully converted all Tooloop-Example into .debs:

- image-slideshow
- kiosk-browser
- video-player
- processing-example (new)

I have also created some packages, that will replace some stuff in the optional installer:

- gstreamer
- processing-3.4
- transparent-cursor

The processing-example is simply a wrapped up example sketch from the processing website. However, it depends on the package processing-3.4 which it will automatically install along when typing:

    sudo apt install tooloop-processing-example

Different Processing versions will also happily co-exist :-)

Now those Debian-Packages are cool, but I will have to provide some information for the settings-server (or future online repo browser). This is where the standard ends and distros like Ubuntu or elementary brew their own stuff. I guess so are we ;-)

Basically we'll have get the .deb on the machine, as well as some meta info and images for the AppCenter.

http://tooloop.org/Documentation/App%20Development/App%20Bundles

Btw. this repository now is local but you can easily turn it into an online one or add another for a future online package store so everything discussed above will be doable.

**Comment by vollstock** [comment]
I will probably wrap up, what I have learned so far (soon) and create a new GithHub repo for those packages. Next step would be branching Tooloop-Settings-Server and Tooloop-OS for the new .deb-based _App Center backend_.
Then we should be good to start on working the new App Center :-)

**Comment by vollstock** [comment]
most of that stuff is being worked on in the appcenter branch now.
It’s not done yet, but I'll close the concept issue.
